orttraining/orttraining/python/training/_ortmodule_utils.py

```diff
@@ -0,0 +1,103 @@
+
+from . import _utils
+
+import onnxruntime
+from onnxruntime.capi.onnxruntime_inference_collection import OrtValue
+from onnxruntime.capi import _pybind_state as C
+
+import torch
+from torch.utils.dlpack import from_dlpack, to_dlpack
+from torch.utils.cpp_extension import load_inline
+
+from enum import IntEnum
+
+class Verbosity(IntEnum):
+    VERBOSE = 0
+    INFO = 1
+    WARNING = 2
+    ERROR = 3
+    FATAL = 4
+
+def _ortvalue_to_torch_tensor(ortvalue):
+    # PyTorch's to_dlpack() uses same config for both torch.bool and torch.uint8,
+    # and convert the config to torch.uint8 tensor duing from_dlpack().
+    # So we need to convert the torch tensor to torch.bool type if OrtValue is bool tensor.
+    torch_tensor = from_dlpack(ortvalue._ortvalue.to_dlpack())
+    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+
+
+def _ortvalue_from_torch_tensor(torch_tensor):
+    return OrtValue(C.OrtValue.from_dlpack(to_dlpack(torch_tensor), torch_tensor.dtype == torch.bool))
+
+def _load_torch_allocator_cpp_extension(verbosity):
+    torch_cuda_allocator_addresses_cpp_source = """
+    #include <torch/extension.h>
+    #include <c10/cuda/CUDACachingAllocator.h>
+    size_t cuda_caching_allocator_raw_alloc_address() {
+        return reinterpret_cast<size_t>(&c10::cuda::CUDACachingAllocator::raw_alloc);
+    }
+    size_t cuda_caching_allocator_raw_delete_address() {
+        return reinterpret_cast<size_t>(&c10::cuda::CUDACachingAllocator::raw_delete);
+    }
+    """
+
+    return load_inline(name='inline_extension', cpp_sources=[torch_cuda_allocator_addresses_cpp_source],
+                       functions=['cuda_caching_allocator_raw_alloc_address',
+                                  'cuda_caching_allocator_raw_delete_address'],
+                       verbose=verbosity < Verbosity.WARNING, with_cuda=True)
+
+def _check_same_device(device, argument_str, *args):
+    '''Check that all tensor arguments in *args reside on the same device as the input device'''
+
+    for arg in args:
+        if arg is not None and isinstance(arg, torch.Tensor):
+            arg_device = torch.device(arg.device)
+            if arg_device != device:
+                raise RuntimeError(
+                    f"{argument_str} found on device {arg_device}, but expected it to be on module device {device}.")
+
+def _create_iobinding(io_binding, inputs, model, device):
+    '''Creates IO binding for a `model` inputs and output'''
+    for idx, value_info in enumerate(model.graph.input):
+        io_binding.bind_ortvalue_input(
+            value_info.name, _ortvalue_from_torch_tensor(inputs[idx]))
+
+    for value_info in model.graph.output:
+        io_binding.bind_output(value_info.name, device.type,
+                               device_id=_utils.get_device_index(device))
+
+def _run_forward(execution_session, onnx_model, device, *inputs, **kwargs):
```

Comment:
Should this be a helper at `ExecutionManager`? Both Inference and Training use it, but in slightly different spots. This is specialized to how c++ module graph is built

Reply:
I moved this method to GraphExecutionManager